### DISCUSSION TITLE
Issue Porting

### DISCUSSION THREAD

**Discussion by meichthys** [TriliumNext/Notes#16] [General]
How do we want to handle / migrate old Trilium issues? Does anyone have experience with this? Is it even possible to migrate the issues automatically (and would we even want to?)

## Update
We are tracking the progress of porting issues/feature requests here:
https://github.com/orgs/TriliumNext/projects/1

If you would like to help out with this project, you will need to be a member of the "Issue Triaging" team.

**Comment by j-tirelli** [comment]
I think there is value in this, it's our best source of feature requests, bug reports, etc..

Preliminary googling says it's possible, maybe even with in-built tools on github

**Comment by eliandoran** [comment]
@meichthys , @j-tirelli ,

I've thought a bit about this when first creating the org. The question is... do we _really_ need to port them?

There are currently:

1. [304 bugs](https://github.com/zadam/trilium/issues?q=is%3Aopen+is%3Aissue+label%3A%22Type%3A+Bug%22+)
2. [408 feature requests](https://github.com/zadam/trilium/issues?q=is%3Aopen+is%3Aissue+-label%3A%22Type%3A+Bug%22+label%3A%22Type%3A+Enhancement%22+)
3. [258 bugs / feature requests / questions](https://github.com/zadam/trilium/issues?q=is%3Aopen+is%3Aissue+-label%3A%22Type%3A+Bug%22+-label%3A%22Type%3A+Enhancement%22+) (uncategorized)

Copying the issues would:

- Cause confusion for the one who created the issue, seeing that it has been copied in another repo (under the name of a bot, most likely).
- Cause duplication between upstream and here, making it difficult to track which features/bugs have been fixed and where.
- Make it difficult for us to triage our own work, where we can use issues to track tasks and see bugs that present only on our fork (e.g. conversion issues).

What I think would be the best approach is to actually help @zadam clean up a bit the issues on his repository, as follows:

- For bug reports, we can have fixes done on our repo and optionally backported upstream. In our commits & PRs we can reference issues from zadam's repo just as well. 
- For feature requests, if we manage to implement them and deliver them, mark the issue as won't fix on @zadam 's repo (since it's in maintenance) and redirect to ours.

Let me know your thoughts.

**Reply by j-tirelli** [reply]
I think this makes sense.

**Reply by meichthys** [reply]
Ok, i can get on board with this, but i think we will still get lots of 'duplication between upstream and here' when people file new issues here that already exist on upstream, but at least with this method, we know the issues here are relevant to us. We can decide to address this as those issues come in (Either triage to find related upstream issue, or just let the new issue stand on it's own).

**Comment by ProfDoof** [comment]
So, I made an issue related to this by accident as I hadn't seen this already. Here's my thought process:

1. zadam's repo is in maintenance mode so the only things we should be possibly be contributing back is bug fixes but realistically, that may introduce far more difficulty as we diverge further from the original code base so it might not be best to try to track them through zadam's repo.
2. There's a non-zero possibility that some feature requests likely fall under a broader umbrella of a feature such as "Switch to X editor" which would then allow for plugins or us to support those features and we should likely group them up and then knock them out that way.
3. You're right that it would make it difficult to track them if they were duplicated so it might be best to coordinate with zadam to transfer the feature requests to this repo so that they are centralized. 
4. Bug reports maybe should be copied to this repository as they could be repaired in either place and maybe should be repaired in both places but tracking where they are fixed and where they aren't probably means that having them only on zadam's repo will make it more difficult. 

So, my suggestion would be to go through the bug reports, feature requests, and questions, and organize them.

- Features should be organized into umbrella features that either summarize how an umbrella feature address that feature or adds future support for that feature through the use of plugins, themes, etc. (Basically a simplified design document of some kind)
- Bug reports should be organized into things that can be addressed during different phases of the current refactoring process or things that need to be considered later or even things that would be addressed by simply implementing a different feature. (Such as switching from ckeditor to a different editor with better support for some thing that is needed to fix the bug)
- Questions should probably be answered on zadam's repository (if someone knows the answer) and possibly added to an FAQ if they are seen repeatedly. 

Also, this reminds me, but we should probably add an issue for either replicating the wiki or providing a docs site for the same stuff.

**Reply by meichthys** [reply]
This sounds really good to me. Especially number 3. I forgot that you can transfer between repositories. That would be a perfect solution for feature requests. I'd be glad to help get started on this, although we would need account permissions from @zadam to migrate feature requests.

**Reply by eliandoran** [reply]
Transferring issues sounds nice. We can already have a look and see what would be nice to transfer/copy in terms of features.

**Reply by CrO2Cl2** [reply]
I think we should make an "Issue Porting" team that can take care of this. I would be more than happy to be one of the people to take on this task, as my coding skills are not the best, but I want to contribute to this project nevertheless

**Reply by meichthys** [reply]
@CrO2Cl2 ive invited you to the issues team.

**Reply by CrO2Cl2** [reply]
how should we organize ourselves @meichthys ? as of now, i think the best way would be to track all the issues that we already had a look at  in a spreadsheet, then aggregate the issues and rewrite them in  TriliumNext's issues tab

**Reply by meichthys** [reply]
Before we re-create any issues in TriliumNext, we should review the existing Trilium issues and make a list of the ones we think are good candidates for moving to TriliumNext. Once we have a list, we can ask Zadam if he would be ok if we transfered those issues out to TriliumNext. This would clean up his repo and move relevant issues to ours. 

Currently i think the plan is to migrate any feature requests to TriliumNext since zadam doesn't have plans to pursue significant feature requests. 

As for normal issues/bugs, we should probably leave those in Trilium but perhaps we can create new issues in our repo that link to the Trilium issues. This would allow us to create a single issue and link to multiple Trilium issues that are similar/relevant (There are lots of duplicate/similar issues in Trilium).

**Reply by CrO2Cl2** [reply]
Sounds great, will start to look through the feature requests and start to aggregate them, how should we communicate? Should we just make another room on matrix and talk through there?

**Reply by meichthys** [reply]
I've created a Project here: https://github.com/orgs/TriliumNext/projects/1

@CrO2Cl2 Is there a trilium matrix?

**Reply by CrO2Cl2** [reply]
afaik we settled on using the matrix room in [this](https://github.com/orgs/TriliumNext/discussions/7) discussion

**Reply by meichthys** [reply]
Ah yes, that droped off my radar. That works for me. We can break out into a Thread if we don't want to clutter up the main chat.

**Reply by maphew** [reply]
> ...we should probably add an issue for either replicating the wiki or providing a docs site for the same stuff.

I was about to act on this and create an issue, but my thoughts aren't well formulated yet, and issues are best kept to things that can be opened and closed. So a discussion thread instead: #62 Trilium hosted wiki?

**Comment by meichthys** [comment]
I've created a project here we can use to track our progress of porting issues: https://github.com/orgs/TriliumNext/projects/1

**Reply by eliandoran** [reply]
@meichthys , the board looks really cool, especially the draft issues. Keep up the good work! :)
When you feel ready, I could drop an email to zadam to see if he would be OK with moving some of these issues.

**Reply by meichthys** [reply]
@eliandoran sounds good 👍
I have permissions on zadams repo to close issues so we've been closing  and cleaning some issues that are clearly fixed or no longer relevant. When we are ready to migrate feature requests, I'll let you know.  We will manually migrate relevant bugs and link to the corresponding issues on Trilium.

I'm looking forward to the beta release!

**Reply by meichthys** [reply]
@eliandoran The first step of the feature-requests migration is complete (A huge thanks to @CrO2Cl2). All the feature requests from Trilium have been grouped/categorized into individual 'draft issues'. If you want to check with zadam to see if he is ok with migrating those feature requests, then we could migrate all feature requests from Trilium, then on our end merge the feature requests by keeping open the most descriptive feature request for each category, and close the others, but link them to the one we kept open.

After migrating all the feature requests, we'll port important issues manually:

> We will Identify (and group) bugs that are important enough to re-create in TriliumNext (These will not be migrated from Trilium, but will need to be manually ported).
> 
> Once we have all relevant bugs categorized, we can convert each category table 'item' to a single TriluimNext issue which will contain link(s) to the original bug issues.
> 
> Ported issues should not include general how-to issues or issues that are very sparse on details/logs/information.

**Reply by eliandoran** [reply]
@zadam , I've sent you an email. We would appreciate your inputs, should you have the time.

**Reply by zadam** [reply]
I'm fine with the transfer.

**Reply by eliandoran** [reply]
@meichthys  , @CrO2Cl2 , let's go. 🚀 
Thank you, @zadam !

**Reply by zadam** [reply]
BTW, don't feel obliged to get my "permission" on such matters. Especially if it might delay things...

**Reply by meichthys** [reply]
@eliandoran I will need write permissions on TriliumNext to make the transfers:
https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository

![Screenshot_20240505-080220.png](https://github.com/TriliumNext/Notes/assets/10717998/b5fc2f1c-8093-4b94-bcc6-863323b48455)

**Reply by eliandoran** [reply]
@meichthys , you should now have full access to the entire org. Let me know if it works for you.

**Reply by meichthys** [reply]
Thanks. I'll get to work on these 👍

**Reply by meichthys** [reply]
@zadam it looks like I'll need more permissions on the Trilium repo as well. I'll need [write access.](https://www.howtogeek.com/devops/how-to-manage-permissions-for-members-in-a-github-organization/)

**Reply by zadam** [reply]
I believe you should already have the write access:

> Collaborators on a personal repository can pull (read) the contents of the repository and push (write) changes to the repository.

> Note: In a private repository, repository owners can only grant write access to collaborators. Collaborators can't have read-only access to repositories owned by a personal account.

https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-user-account-settings/permission-levels-for-a-personal-account-repository#collaborator-access-for-a-repository-owned-by-a-personal-account

**Reply by meichthys** [reply]
@zadam  Apparently i was wrong about only needing write access. 
It looks like repository ownership is required too:

> Note: You can only transfer issues between repositories **`owned by the same user or organization`** account. A private repository issue cannot be transferred to a public repository.

If you prefer, we can just create our own copies of the feature requests on the Trilium repository and then link back to the original feature requests.

**Reply by CrO2Cl2** [reply]
it would be nice for me to close the original feature requests, as that would make my life easier when I later do another pass over the issues to see if something important slipped through the cracks.

**Reply by meichthys** [reply]
Agreed, but that'll be up to Zadam since we would need permissions to do the transfers, or he would have to do them himself.

**Reply by CrO2Cl2** [reply]
I wass thinking, couldn't we just make the drafts from our project, make them into issues? this would mean that they are already aggregated, instead of porting over eevery issue that has to do with that topic? and then if we get the permission close the issues in Zadams repo?

**Reply by meichthys** [reply]
Yes, that's what i was thinking as an alternative, but we would loose the issue descriptions and would need to rely on links to closed issues on the Trilium Repo.
I think my preference would be to transfer the issues, pick the best issue to leave open for each 'category', and then close and link the other related issues to it. This way the issue discussion wouldn't be buried in a link.

**Reply by CrO2Cl2** [reply]
I do understand where you come from, but on lots and lots of issues that suggest a pretty good idea, there is absolutely no more information. Eg the really popular request of excel like functionality. the name says it all. on most of them there is absolutely no discussion, just that. some of the issues have slightly different ideas, but it is mostly the same. perhaps we could transfer some where there is a lot of info in the comments, and just remake the ones where no to little information is given? that is if we get the transfer permission

**Reply by meichthys** [reply]
I can get on board with that as it would make the migration a bit easier for everyone involved. 
Once we create the issues, I can go back and close the originals in the Trilium repo since i do have permissions to do that for Zadam. One downside would be that the original user who filed the issues would not be redirected to our repo. Their issue would just be closed, but at least there would be a link there that they could follow to the new issue (FR).

**Reply by CrO2Cl2** [reply]
I think that if we write a message in the original issue that it has been moved, that should be more than enough. If the issue is very old and the OP does not care anymore, the he can just ignore the message instead of getting lots of pings 3 years later. if the OP is still interested, he can just click on the link and follow the conversation

**Reply by meichthys** [reply]
That sounds good to me. Let's wait for zadam's response to see if he has a preference. If not, then we'll move forward with converting the draft issues into feature requests 👍

**Reply by eliandoran** [reply]
@meichthys , I think we should not bother @zadam with asking him to transfer the issues on his own.
We need to find another way. 
I think duplicating the description + link to the old issue + closing the old issue (as per @CrO2Cl2 ) would be the best solution given the situation.

**Reply by meichthys** [reply]
> @meichthys , I think we should not bother @zadam with asking him to transfer the issues on his own.

I wasn't expecting zadam to do the transfers himself. I only was waiting since zadam thought it would be helpful to him if we transfered them out (perhaps less issues to search through). If we had owner permission for an hour or so, we should be able to do the transfer ourselves. It would probably be a more seamless experience for the original feature-request reporters if they saw their issue migrated to our repo - that way they would be able to follow the issue and they would realize that future development of feature-requests would be happening on TriliumNext. Closing the issue isn't a bad solution, but it does obscure the original reports from our view (unless we manually copy over the text of the feature requests or rely on links).

All that being said, I'm fine with moving forward with manually recreating the feature requests in our repo (converting our existing project 'draft issues' into actual issues).

**Reply by eliandoran** [reply]
@meichthys , I've thought about temporary access as well and did discuss this with zadam but it does not seem to be possible for personal repos, only within an organization.

Case in which we're stuck either with a script of some kind to automate the process or manually doing it.

https://stackoverflow.com/questions/64478636/how-to-give-admin-access-to-private-repository-when-there-is-no-role-options

**Reply by meichthys** [reply]
Ah.. ok. in that case, let's get this thing moving!
I'll get started today in converting our existing 'draft' feature request categories over to issues. Where relevant, i'll copy over the text from the best feature request on the original repo. 

Once the manual 'transfers' are complete, I'll go ahead and close out the issue on Trilium.

**Reply by eliandoran** [reply]
@meichthys , sounds great.
I've created a label called `ported-issues` which we could assign to the new issues in order to be able to easily identify them from our own issues.

**Reply by meichthys** [reply]
Good idea. Will do 👍

**Comment by CrO2Cl2** [comment]
I have finished aggregating all the issues to the Project

**Reply by meichthys** [reply]
Wow great work!
I will review soon 👍
